### ISSUE TITLE
Directories in custom_build deps not working

### ISSUE THREAD

**Issue by fgblomqvist** [tilt-dev/tilt#4605]
### Expected Behavior
`custom_build(... deps=["src/hello-world"])` should lead to that image getting rebuilt whenever any files in `src/hello-world` change.

### Current Behavior
Changes to underlying files/folders do not trigger changes. Adding a specific file to the deps list works: `deps=["src/hello-world/foo.txt"]` but directories don't seem to work.

### Steps to Reproduce

1. Add a `custom_build` call to your tilt file where you pass a directory to the `deps` arg (as an array of course)
2. Change a file inside the directory
3. See how Tilt does not rebuild that image

#### `tilt doctor` Output

```
$ tilt doctor
Tilt: v0.20.5, built 2021-05-28
System: linux-amd64
---
Docker
- Host: [default]
- Version: 1.41
- Builder: 2
---
Kubernetes
- Env: kind-0.6+
- Context: kind-kind
- Cluster Name: kind-kind
- Namespace: default
- Container Runtime: containerd
- Version: v1.15.12
- Cluster Local Registry: {Host:localhost:5000 hostFromCluster: SingleName:}
---
Thanks for seeing the Tilt Doctor!
Please send the info above when filing bug reports. 💗

The info below helps us understand how you're using Tilt so we can improve,
but is not required to ask for help.
---
Analytics Settings
--> (These results reflect your personal opt in/out status and may be overridden by an `analytics_settings` call in your Tiltfile)
- User Mode: opt-in
- Machine: fdc1841e6eb329ae14cfb4a12a516840
- Repo: GERmNS3VTepO9fN6OZ7L1g==
...
```

#### About Your Use Case

Just the usual stuff, trying to get reactive rebuilds working.

**Comment by fgblomqvist** [comment]
I'm not 100% sure this is a bug, but since there's no "Discussions" section on this repo I figured I'd file it as one for now.

Worth noting is that doing `deps=["."]` works.

**Comment by milas** [comment]
Are you seeing this consistently? I'm not able to repro this on Ubuntu 20.04.

Is there anything unusual about your setup, e.g. is this on a remote filesystem like NFS? Does the issue persist across reboots?

Can you run `sysctl fs.inotify` and post the output here? You might need to increase a limit. Tilt does try to surface this in the logs, but it might not be perfect. Do you see any errors similar to `failed to create fsnotify watcher: too many open files` in Tilt?

**Comment by nicks** [comment]
I tried to reproduce this in one of the example repos, but it worked ok for me:
https://github.com/tilt-dev/tilt-example-html/blob/nicks/issue4605/0-base/Tiltfile

@fgblomqvist I wrote this blog post on how to use `tilt get filewatch` and `tilt describe filewatch` to inspect what files Tilt is watching and test when it picks up a change, maybe it will help? https://blog.tilt.dev/2021/05/07/eyes-on-the-watchers.html

**Comment by fgblomqvist** [comment]
@milas Yes, so far it doesn't work at any time. If I use the `docker_build` stuff it works, but not with the `custom_build`.
The only unusual part of the setup is that everything (including Tilt) runs on a remote server. Though everything runs there (including Tilt), so it's as if I was running things locally. I have not tried rebooting yet but it's something I can do if @nicks suggestions don't lead anywhere.

I haven't seen what you wrote in the logs, but the output of that command is:
```
fs.inotify.max_queued_events = 16384
fs.inotify.max_user_instances = 128
fs.inotify.max_user_watches = 524288
```

@nicks Thanks for the suggestions (great blog post). This info API is really cool/useful.
Here's some relevant data (found the filewatch for the image in question so that part seems to work):
```yaml
apiVersion: tilt.dev/v1alpha1
kind: FileWatch
metadata:
  <snip>
spec:
  ignores:
  - basePath: .git
  - basePath: /home/fblomqvist/X/.git
  - basePath: /home/fblomqvist/X/foo/bar/baz
    patterns:
    - '*'
    - .logs
    - .DS_Store
    - '!.env'
    - '!src'
    - '!lib'
    - '!dev'
    - '!env'
    - '!.chart'
    - '!.gitlab'
    - '!.cicd'
    - '**/.terraform/*'
    - '**/.terragrunt-cache/*'
    - '*.tfstate'
    - '*.tfstate.*'
    - crash.log
    - '*.tfvars'
    - ~*.tf
    - ~*.hcl
    - override.tf
    - override.tf.json
    - '*_override.tf'
    - '*_override.tf.json'
    - .terraformrc
    - terraform.rc
    - tilt_modules
    - node_modules
  watchedPaths:
  - /home/fblomqvist/X/foo/bar/baz
status:
  lastEventTime: null
  monitorStartTime: "2021-06-07T21:21:07.915892Z"
```

Could the issue be that Tilt reads the .gitignore at the repo root (it's a monorepo) and then uses that inside the specific project (foo/bar/baz). That doesn't feel like the correct behavior? The patterns defined in the gitignore are relevant when evaluated from the root of the repo, not when evaluated inside various folders (not saying it's a great gitignore, suffers from some legacy stuff, but that's something to tackle another day 😅 ). Though, that being said, the only gitignore pattern that could ruin things would be `*` but that should only be on a single level. Even if I modify a file in `foo/bar/baz/a/b/c` it does not trigger a reload.

The Tiltfile that contains this `custom_build` sits inside `foo/` and is loaded into the main Tiltfile using `load_dynamic`.

**Comment by nicks** [comment]
@fgblomqvist good find! As far as I can tell, Tilt doesn't read your gitignore. Is it possible those patterns are coming from your dockerignore? let me try to repro with those patterns

**Comment by nicks** [comment]
ya, I can confirm that with those ignore patterns (in a .dockerignore), it's expected that Tilt would ignore changes under `foo/bar/baz/a/b/c`.

maybe the real bug here is that it's reading dockerignore patterns from someplace unexpected, or it's evaluating them relative to an unexpected base path?

**Comment by fgblomqvist** [comment]
Walked up the dir tree and yeah there are 2 dockerignores (that are essentially copies of the gitignores). One is in the same dir as the root of the repo (which is also where the main Tiltfile lives), and one is in a dir between the root of the repo and the root of the project I'm watching.

E.g. per the example above, one sits in `/home/fblomqvist/X` and one sits in `/home/fblomqvist/X/foo`. So yeah, I think you might be right about Tilt evaluation them in the wrong dir maybe? 🙂

**Comment by nicks** [comment]
OK, so there's definitely something wrong with both
1) which .dockerignore files Tilt reads, and
2) how it determines the base path of those dockerignores

But I'm not totally sure what the right behavior should be.

My instinct is that we should be "maximally" consistent with `docker build` semantics, which means that:

```
docker_build(imageName, './path/to/context')
custom_build(
  imageName,
  'docker build -t $EXPECTED_REF frontend',
  ['./path/to/context'],
)
```

should be equivalent. So in this case, we'd treat the first dir in "deps" as special, and read "./path/to/context/.dockerignore" (and resolve it relative to "./path/to/context")

**Comment by fgblomqvist** [comment]
Wouldn't it read the dockerignore in every dep dir (if one exists) and then evaluate them in their respective dirs (only)? Or is that what you meant?

**Comment by nicks** [comment]
@fgblomqvist hmmm...I'm not sure! I could probably be convinced either way...

**Comment by fgblomqvist** [comment]
Well, in that case I vote for what I said 😄 . But feel free to think about it and probably consult with the other maintainers.

**Comment by nicks** [comment]
this should be fixed in 0.20.7

**Comment by fgblomqvist** [comment]
Verified that it works now, thanks!
